logictest: give larger disk limit to clusters in big tests

yuzefovich opened the pull request:
This commit adds an option to specify the temp storage limit for
test clusters created to run the logic tests and uses that option to
increase the limit when running the big test (sqllitelogic).

Fixes: #48283.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/47381)
<!-- Reviewable:end -->

asubiotto reviewed (changes requested):
Is the thinking here that we run into a disk limit and return `already spilled to disk`? If that's the case, I think we're probably using a `OutOfMemory` error code when hitting a memory limit when we should probably be returning `DiskFull`.

> Reviewed 1 of 5 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47381#-:-M4nzt4WAhIsiBLcUpsj:b-99boie)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @jordanlewis, and @yuzefovich)

___
*[pkg/base/test_server_args.go, line 154 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47381#-M4nznYa7JJqHNddty8m:-M4nznYa7JJqHNddty8n:bvcg25w) ([raw file](https://github.com/cockroachdb/cockroach/blob/ada69963b92bd2a9930bda185f1d7c4dd417d536/pkg/base/test_server_args.go#L154)):*
> ```Go
> func DefaultTestTempStorageConfig(st *cluster.Settings) TempStorageConfig {
> 	var maxSizeBytes int64 = DefaultInMemTempStorageMaxSizeBytes
> 	return DefaultTestTempStorageConfigWithSize(st, maxSizeBytes)
> ```

nit: pass in `DefaultInMemTempStorageMaxSizeBytes` directly.
___
*[pkg/sql/logictest/logic.go, line 1126 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47381#-M4o0T0m13CObJfKMSIs:-M4o0T0m13CObJfKMSIt:b-nnwrbe) ([raw file](https://github.com/cockroachdb/cockroach/blob/ada69963b92bd2a9930bda185f1d7c4dd417d536/pkg/sql/logictest/logic.go#L1126)):*
> ```Go
> }
> 
> func (t *logicTest) setup(cfg testClusterConfig, serverArgs TestServerArgs) {
> ```

I think it's a bit confusing to call this `TestServerArgs`, considering we use `base.TestServerArgs`. Can we add what we need to `testClusterConfig`?


<!-- Sent from Reviewable.io -->

asubiotto commented:
OK I think I understand the issue. The problem is that our disk monitor is a `MemoryResource` in the in-memory engine case, which means it returns a memory, rather than a disk error when full (which causes "already using disk"). We just need to change `mon.MemoryResource` to `mon.DiskResource` in `TempStorageConfigFromEnv`.

yuzefovich reviewed (commented):
I agree that we were using incorrect resource type for in-memory temp storage, but I don't think this will solve the "already using disk" problem. We should discuss it offline.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47381#-:-M4zMcsy4u9aLypwyEDD:b-ctroi5)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @jordanlewis)

___
*[pkg/base/test_server_args.go, line 154 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47381#-M4nznYa7JJqHNddty8m:-M4zM_u88xO4Qbt_xNpp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ada69963b92bd2a9930bda185f1d7c4dd417d536/pkg/base/test_server_args.go#L154)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

nit: pass in `DefaultInMemTempStorageMaxSizeBytes` directly.
</blockquote></details>

Done.
___
*[pkg/sql/logictest/logic.go, line 1126 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47381#-M4o0T0m13CObJfKMSIs:-M4oizUq0RLMbSTcycET:b-ti1sna) ([raw file](https://github.com/cockroachdb/cockroach/blob/ada69963b92bd2a9930bda185f1d7c4dd417d536/pkg/sql/logictest/logic.go#L1126)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I think it's a bit confusing to call this `TestServerArgs`, considering we use `base.TestServerArgs`. Can we add what we need to `testClusterConfig`?
</blockquote></details>

I don't think that we can - the goal here is to be able to call `RunLogicTest` with custom limits, at that point we don't have access to `testClusterConfig` because it is unexported and the caller of `RunLogicTest` doesn't need to know about different cluster configs that we support. I think it should be a separate argument and a separate struct because it is "larger" configuration options.

I agree that there is a slight confusion with the naming, but `logictest.TestServerArgs` is a "subset" of `base.TestServerArgs` and I think it sends a message that using the former will influence what values are used when instantiating the latter. I'm open to other naming suggestions though.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
Kicked off big test manually and saw this:
```
6187 [20:20:56] :› ›  [run] {"Time":"2020-04-15T16:20:56.64659207-04:00","Action":"output","Package":"github.com/cockroachdb/cockroach/pkg/sql/logictest","Test":"TestSqlLiteLogic/lo      cal/slt_good_30.test#01","Output":"panic: non-values node passed as fixed value to zigzag join [recovered]\n"}^M
6188 [20:20:56] :› ›  [run] {"Time":"2020-04-15T16:20:56.646648327-04:00","Action":"output","Package":"github.com/cockroachdb/cockroach/pkg/sql/logictest","Test":"TestSqlLiteLogic/l      ocal/slt_good_30.test#01","Output":"\tpanic: non-values node passed as fixed value to zigzag join [recovered]\n"}^M
6189 [20:20:56] :› ›  [run] {"Time":"2020-04-15T16:20:56.64665632-04:00","Action":"output","Package":"github.com/cockroachdb/cockroach/pkg/sql/logictest","Test":"TestSqlLiteLogic/lo      cal/slt_good_30.test#01","Output":"\tpanic: panic while executing 1 statements: SELECT _ FROM _ WHERE (((_ IS _) AND (_ \u003c _)) AND (_ \u003e= _)) OR ((_ IS _) AND (_ IS _))       ORDER BY _ DESC; caused by non-values node passed as fixed value to zigzag join\n"}^M
```
So temporarily added another commit to disable parallelism in big test.

yuzefovich commented:
Hm, I don't understand this - I hit `fatal out of memory` error when running big test on CI with `-show-sql` when I explicitly specify the default values for SQL root memory and temp storage limit. Any idea what could be the problem?

asubiotto reviewed (commented):
> I agree that we were using incorrect resource type for in-memory temp storage, but I don't think this will solve the "already using disk" problem.

Why not? When trying to add a row and hitting the disk limit we'll return an out of memory instead of out of disk error, since we're already "using disk" at this point we return "already using disk"

No idea regarding the `fatal out of memory`. Are you saying that we're running into a fatal OOM when setting these limits higher?

> Reviewed 2 of 5 files at r1, 4 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47381#-:-M56_eAk2hbQoAhnk0FT:b-4rb6vw)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @jordanlewis)



<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> Why not? When trying to add a row and hitting the disk limit we'll return an out of memory instead of out of disk error, since we're already "using disk" at this point we return "already using disk"

I guess it depends on what we consider as "solving" this problem - if we switch the resource, I expect that we'll be getting "disk full" error rather than our custom "already using disk". If this is considered the desired outcome, then I agree, updating the resource type solves the problem. When I said that it would not solve the problem I meant that the query is still expected to fail. (Unless I'm missing something.)

Turns out we're running into a fatal OOM even on master, when keeping the limits unchanged (as well as on this branch with higher limits). I haven't looked into that deeper.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47381#-:-M57o2mH-CApmmKER_wg:be2j8jh)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @jordanlewis)



<!-- Sent from Reviewable.io -->

asubiotto commented:
Fair enough. The `already using disk` is one issue but we need to investigate these `fatal out of memory` despite the limit increases. Do you want to merge a fix for `already using disk` and investigate this OOM/increase disk limits separately?

yuzefovich commented:
Good point, extracted into a separate PR.

asubiotto commented:
Now that #47976 is in, is this RFAL?

yuzefovich commented:
No. Now zigzan join related panic occurs when running big test (which needs to be investigated), but I haven't seem OOMs, so it's possible that this PR is no longer necessary, and OOMs were caused solely by #47976.

yuzefovich commented:
We've hit a temp storage limit a few times, so I think it's still worth bumping it up. I removed the knob for sql root memory limit though.

Kicked off test run: https://teamcity.cockroachdb.com/viewLog.html?buildId=1922843&buildTypeId=Cockroach_SqlLogicTests&fromExperimentalUI=true, let's see what happens.

yuzefovich commented:
Ok, so bumping up temp storage limit, increasing the timeout from 24 hours to 48 hours and disabling batch size randomization made the test pass.

Let's try another time with batch size always 1. The link: https://teamcity.cockroachdb.com/viewLog.html?buildId=1925243&buildTypeId=Cockroach_SqlLogicTests&fromExperimentalUI=true

asubiotto commented:
Interesting. The test completed in 12h. Is the bump to 48h necessary?

yuzefovich commented:
I don't know yet. 12 hours was with disabled batch size randomization, so I kicked off the run with batch size always set to 1 in order to see what is the upper bound.

yuzefovich commented:
Oh, the test completed with fixed batch size of 1 in about 12 hours, so it does seem like the timeout bump is not needed.

I wonder whether without bumping the temp storage limit we hit "disk full" error in the vectorized engine and somehow get "stalled", and the query is left hanging, which is the actual reason of timing out.

yuzefovich commented:
Let's try that theory - I actually decreased temp storage limit to 32 MiB and slightly decreased timeout to 16 hours, [link](https://teamcity.cockroachdb.com/viewLog.html?buildId=1927681&).

yuzefovich commented:
The test timed out, and it looks like there is a deadlock in draining of the hash router:
```
 16747 [11:26:16] :› ›  [run] {"Time":"2020-05-08T07:26:14.110143437-04:00","Action":"output","Package":"github.com/cockroachdb/cockroach/pkg/sql/logictest","Test":"TestSqlLiteLogic/fak       edist/slt_good_4.test#08","Output":"goroutine 417824690 [semacquire, 702 minutes]:\n"}^M
 16748 [11:26:16] :› ›  [run] {"Time":"2020-05-08T07:26:14.110149875-04:00","Action":"output","Package":"github.com/cockroachdb/cockroach/pkg/sql/logictest","Test":"TestSqlLiteLogic/fak       edist/slt_good_4.test#08","Output":"sync.runtime_SemacquireMutex(0xc037c80d54, 0xc0295ef300, 0x1)\n"}^M
 16749 [11:26:16] :› ›  [run] {"Time":"2020-05-08T07:26:14.110157079-04:00","Action":"output","Package":"github.com/cockroachdb/cockroach/pkg/sql/logictest","Test":"TestSqlLiteLogic/fak       edist/slt_good_4.test#08","Output":"\t/usr/local/go/src/runtime/sema.go:71 +0x47\n"}^M
 16750 [11:26:16] :› ›  [run] {"Time":"2020-05-08T07:26:14.110166511-04:00","Action":"output","Package":"github.com/cockroachdb/cockroach/pkg/sql/logictest","Test":"TestSqlLiteLogic/fak       edist/slt_good_4.test#08","Output":"sync.(*Mutex).lockSlow(0xc037c80d50)\n"}^M
 16751 [11:26:16] :› ›  [run] {"Time":"2020-05-08T07:26:14.110172913-04:00","Action":"output","Package":"github.com/cockroachdb/cockroach/pkg/sql/logictest","Test":"TestSqlLiteLogic/fak       edist/slt_good_4.test#08","Output":"\t/usr/local/go/src/sync/mutex.go:138 +0xfc\n"}^M
 16752 [11:26:16] :› ›  [run] {"Time":"2020-05-08T07:26:14.110179082-04:00","Action":"output","Package":"github.com/cockroachdb/cockroach/pkg/sql/logictest","Test":"TestSqlLiteLogic/fak       edist/slt_good_4.test#08","Output":"sync.(*Mutex).Lock(...)\n"}^M
 16753 [11:26:16] :› ›  [run] {"Time":"2020-05-08T07:26:14.110185293-04:00","Action":"output","Package":"github.com/cockroachdb/cockroach/pkg/sql/logictest","Test":"TestSqlLiteLogic/fak       edist/slt_good_4.test#08","Output":"\t/usr/local/go/src/sync/mutex.go:81\n"}^M
 16754 [11:26:16] :› ›  [run] {"Time":"2020-05-08T07:26:14.11019165-04:00","Action":"output","Package":"github.com/cockroachdb/cockroach/pkg/sql/logictest","Test":"TestSqlLiteLogic/fake       dist/slt_good_4.test#08","Output":"github.com/cockroachdb/cockroach/pkg/sql/colexec.(*routerOutputOp).drain(0xc037c80d20, 0x0, 0x0, 0x0)\n"}^M
 16755 [11:26:16] :› ›  [run] {"Time":"2020-05-08T07:26:14.110200386-04:00","Action":"output","Package":"github.com/cockroachdb/cockroach/pkg/sql/logictest","Test":"TestSqlLiteLogic/fak       edist/slt_good_4.test#08","Output":"\t/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/routers.go:377 +0x18b\n"}^M
 16756 [11:26:16] :› ›  [run] {"Time":"2020-05-08T07:26:14.110207131-04:00","Action":"output","Package":"github.com/cockroachdb/cockroach/pkg/sql/logictest","Test":"TestSqlLiteLogic/fak       edist/slt_good_4.test#08","Output":"github.com/cockroachdb/cockroach/pkg/sql/colexec.(*HashRouter).DrainMeta(0xc07af27290, 0x440e580, 0xc0629e0bc0, 0x0, 0x0, 0x0)\n"}^M
 16757 [11:26:16] :› ›  [run] {"Time":"2020-05-08T07:26:14.110214153-04:00","Action":"output","Package":"github.com/cockroachdb/cockroach/pkg/sql/logictest","Test":"TestSqlLiteLogic/fak       edist/slt_good_4.test#08","Output":"\t/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/routers.go:622 +0xf0\n"}^M
 16758 [11:26:16] :› ›  [run] {"Time":"2020-05-08T07:26:14.110229538-04:00","Action":"output","Package":"github.com/cockroachdb/cockroach/pkg/sql/logictest","Test":"TestSqlLiteLogic/fak       edist/slt_good_4.test#08","Output":"github.com/cockroachdb/cockroach/pkg/sql/colexec.NewMaterializer.func1(0x440e580, 0xc0629e0bc0, 0x0, 0x0, 0x0)\n"}^M
 16759 [11:26:16] :› ›  [run] {"Time":"2020-05-08T07:26:14.110241143-04:00","Action":"output","Package":"github.com/cockroachdb/cockroach/pkg/sql/logictest","Test":"TestSqlLiteLogic/fak       edist/slt_good_4.test#08","Output":"\t/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:107 +0xc4\n"}^M
```
Will dig into it.

yuzefovich committed:
logictest: give larger disk limit to clusters in big tests

This commit adds an option to specify the temp storage limit for
test clusters created to run the logic tests and uses that option to
increase the limit when running the big test (sqllitelogic).

Release note: None

asubiotto commented:
My suspicion of what's happening here is that we're mistakenly calling `reset` on a router output op from a disk spilling operator. This locks a mutex and runs into a budget exceeded panic, which unwinds the stack but never `Unlock`s the mutex (it's not unlocked in a defer). I don't think we should ever be calling `reset` on a router output op, that was only ever meant for benchmarks. Pushed a change to this branch and rerunning the sql logic tests [here](https://teamcity.cockroachdb.com/viewLog.html?buildId=1942984&tab=queuedBuildOverviewTab)

asubiotto commented:
We're still assuming that we'll be able to hit this deadlock with a timeout of 8h and a batch size of 1, correct?

yuzefovich commented:
Yeah, I believe with low temp storage limit 8 hours is more than sufficient to run into the deadlock, however, I'm not sure whether if the deadlock doesn't occur that we don't hit the timeout.

asubiotto commented:
OK. I guess we should see how this run goes. We'll have to determine what kind of timeout we hit. I think I want to get these changes in on master regardless so we could merge that and see how the the sqlite logic tests fare over the weekend.

asubiotto commented:
The run hit the 8h timeout but it doesn't look like it hit the hashrouter deadlock. I'll create a PR for the hashrouter changes and see how the runs fare over the weekend.

asubiotto commented:
The PR to fix the deadlock is merged. Given that there have been a bunch of failures due to this disk limit I think we should go ahead with this PR and bump it. Is this RFAL?

yuzefovich commented:
RFAL.

asubiotto reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 3 of 3 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47381#-:-M7hpV4j2x7G2JtJhgKM:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto, @jordanlewis, and @yuzefovich)

___
*[pkg/sql/logictest/logic_test.go, line 111 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/47381#-M7hpFvbBwYIEYDiEwVp:-M7hpFvc4wlLtb3oYI--:bgr5y43) ([raw file](https://github.com/cockroachdb/cockroach/blob/c94d33ef440f309e0410dffe697cd6fe3724f7da/pkg/sql/logictest/logic_test.go#L111)):*
> ```Go
> 	// so we give them larger temp storage limit than other logic tests get.
> 	serverArgs := TestServerArgs{
> 		tempStorageDiskLimit: 1 << 29, // 512 MiB
> ```

I personally prefer `512 << 20` because in my mind I see `20` and I think `MB` so it reads more easily.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47381#-:-M7hwSaf88dHp3SHIBoi:b-oduykh)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto and @jordanlewis)

___
*[pkg/sql/logictest/logic_test.go, line 111 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/47381#-M7hpFvbBwYIEYDiEwVp:-M7hwU3O3VBsyVrTEn7f:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c94d33ef440f309e0410dffe697cd6fe3724f7da/pkg/sql/logictest/logic_test.go#L111)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I personally prefer `512 << 20` because in my mind I see `20` and I think `MB` so it reads more easily.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
___
*[pkg/sql/logictest/logic_test.go, line 111 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/47381#-M7hpFvbBwYIEYDiEwVp:-M7hyiRZ0__VppZ2BtL5:b-etqidc) ([raw file](https://github.com/cockroachdb/cockroach/blob/c94d33ef440f309e0410dffe697cd6fe3724f7da/pkg/sql/logictest/logic_test.go#L111)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Done.
</blockquote></details>

512 * 1024 * 1024 is even better :)


<!-- Sent from Reviewable.io -->
